synapse/http/matrixfederationclient.py

```diff
@@ -519,13 +523,26 @@ def put_json(self, destination, path, args={}, data={},
             json=data,
         )
 
-        response = yield self._send_request(
-            request,
-            long_retries=long_retries,
-            timeout=timeout,
-            ignore_backoff=ignore_backoff,
-            backoff_on_404=backoff_on_404,
-        )
+        send_request_args = {
+            "request": request,
+            "long_retries": long_retries,
+            "timeout": timeout,
+            "ignore_backoff": ignore_backoff,
+            # Do not backoff on the initial request if we're trying with trailing slashes
+            # Otherwise we may end up waiting to contact a server that is actually up
+            "backoff_on_404": False if try_trailing_slash_on_404 else backoff_on_404,
+        }
+
+        response = yield self._send_request(**send_request_args)
+
+        # If enabled, retry with a trailing slash if we received a 404
+        if try_trailing_slash_on_404 and response.code == 404:
```

Comment:
you seem to have missed the 400 case here, which brings me to:

how about implementing this with a wrapper which you can use in both cases?

```
def _send_request_with_optional_trailing_slash(request, try_trailing_slash_on_404=False, **kwargs):
    self._send_request(request, **kwargs)
    if that_worked:
        return
    fiddle_with(request)
    self._send_request(request, **kwargs)
```